### ISSUE TITLE
Best way to override controller actions?

### ISSUE THREAD

**Issue by hegwin** [JSONAPI-Resources/jsonapi-resources#921]
I have an `Address` model with following codes:
```ruby
class Address < ActiveRecord::Base

  def update_or_clone(attributes)
    if any_orders?
      object_copy = dup
      object_copy.assign_attributes(attributes)
      Address.transaction do
        update_attribute(:is_discarded, true)
        object_copy.save
      end
    else
      update_attribute(attributes)
    end
end
```

And I have a jsonapi controller, I need to override the `#update` action and use `update_or_clone` instead of normal `update_attributes`.

```ruby
module Api
  module V1
    class AddressesController < JSONAPI::ResourceController
      # TODO
      def update
        logger.info params[:data][:attributes]

        @address = Address.find(params[:id])
        @address.update_or_clone(params[:data][:attributes])

        render json: {}, status: :ok
      end

      def make_default
        @address = Address.find(params[:id])
        @address.make_default!
        render json: @address, status: :ok
      end

      def context
        { current_user: guardian.user }
      end
    end
  end
end
```

If I do this, I won't work. The attributes are not well formatted. My parameters are like these, something like `address-line2` are not transferred to `address_line2`
```
 {"data"=>{"type"=>"addresses", "id"=>"379", "attributes"=>{"name"=>"Fitzwilliam Darcy", "address-line1"=>"Pemberley", "address-line2"=>"Test Line2", "city"=>"Lambton", "state"=>"UX", "zip"=>"12306", "country"=>"EN", "shipping"=>true, "billing"=>false}, "relationships"=>{"consumer"=>{"data"=>{"id"=>624, "type"=>"consumers"}}}}, "id"=>"379"}
{"name"=>"Fitzwilliam Darcy", "address-line1"=>"Pemberley", "address-line2"=>"Test Line2", "city"=>"Lambton", "state"=>"UX", "zip"=>"12306", "country"=>"EN", "shipping"=>true, "billing"=>false}
```

How can I keep the original features provided by jsonapi-resources (1. Params format, 2. Relationship), and override `#update` action as well?
I know that I can generate a new hash and replace the keys. However, I don't think this is a very good way. And I can't work with relations in this way.

**Comment by lgebhardt** [comment]
Because the update process is happening in the processor and the resource you don't want to try and put the logic in the controller like you would in a standard rails app. I would create a custom `AddressProcessor` and override the `replace_fields` method, see http://jsonapi-resources.com/v0.8/guide/operation_processors.html.

```ruby
class AddressProcessor < JSONAPI::Processor
    def replace_fields
      resource_id = params[:resource_id]
      data = params[:data]

      resource = resource_klass.find_by_key(resource_id, context: context)

      # Put your clone logic here and set resource to the cloned address if need be
      
      result = resource.replace_fields(data)

      return JSONAPI::ResourceOperationResult.new(result == :completed ? :ok : :accepted, resource)
    end
end
```

**Comment by hegwin** [comment]
@lgebhardt Thank you.

At first I found I didn't have `JSONAPI::Processor` because I was using `'jsonapi-resources' '0.7.0'`. Thus, I updated `jsonapi-resources` to `0.8.1` and met three deprecation warning about `sprockets`.  Then I locked the version of sprockets to 3.6.3 instead of 3.7.0 to void the deprecation.

And then I got some inspiration from your codes. I looked into the codes for `jsonapi/resources` and I found the definition of [_replace_fields(field_data)](https://github.com/cerebris/jsonapi-resources/blob/master/lib/jsonapi/resource.rb#L375). I think I can overwrite this method with smallest change to achieve what I want.
